src/AppInstallerCommonCore/Downloader.cpp

```diff
@@ -171,51 +171,64 @@ namespace AppInstaller::Utility
-                FILE_ATTRIBUTE_NORMAL, /*dwFlagsAndAttributes*/
-                NULL /*hTemplateFile*/) };
-
-            THROW_LAST_ERROR_IF(fileHandle.get() == INVALID_HANDLE_VALUE);
-
-            wchar_t fileSystemName[MAX_PATH];
-            THROW_LAST_ERROR_IF(!GetVolumeInformationByHandleW(
-                fileHandle.get(), /*hFile*/
-                NULL, /*lpVolumeNameBuffer*/
-                0, /*nVolumeNameSize*/
-                NULL, /*lpVolumeSerialNumber*/
-                NULL, /*lpMaximumComponentLength*/
-                NULL, /*lpFileSystemFlags*/
-                fileSystemName, /*lpFileSystemNameBuffer*/
-                MAX_PATH /*nFileSystemNameSize*/));
-
-            if (_wcsicmp(fileSystemName, L"NTFS") != 0)
-            {
-                AICLI_LOG(Core, Info, << "File system is not NTFS. Skipped applying motw");
-                return;
-            }
+            AICLI_LOG(Core, Info, << "File system is not NTFS. Skipped applying motw");
+            return;
         }
 
-        // Zone Identifier stream name
-        // https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-fscc/6e3f7352-d11c-4d76-8c39-2516a9df36e8
-        std::filesystem::path motwPath(filePath);
-        motwPath += L":Zone.Identifier:$DATA";
+        Microsoft::WRL::ComPtr<IZoneIdentifier> zoneIdentifier;
+        THROW_IF_FAILED(CoCreateInstance(CLSID_PersistentZoneIdentifier, nullptr, CLSCTX_INPROC_SERVER, IID_PPV_ARGS(&zoneIdentifier)));
+        THROW_IF_FAILED(zoneIdentifier->SetId(zone));
 
-        // Apply mark of the web. ZoneId 3 means downloaded from internet.
-        std::ofstream motwStream(motwPath);
-        motwStream << "[ZoneTransfer]" << std::endl;
-        motwStream << "ZoneId=3" << std::endl;
+        Microsoft::WRL::ComPtr<IPersistFile> persistFile;
+        THROW_IF_FAILED(zoneIdentifier.As(&persistFile));
+        THROW_IF_FAILED(persistFile->Save(filePath.c_str(), TRUE));
 
         AICLI_LOG(Core, Info, << "Finished applying motw");
     }
+
+    void ApplyMotwUsingIAttachmentExecuteIfApplicable(const std::filesystem::path& filePath, const std::string& source)
+    {
+        AICLI_LOG(Core, Info, << "Started applying motw using IAttachmentExecute to " << filePath);
+
+        if (!IsNTFS(filePath))
+        {
+            AICLI_LOG(Core, Info, << "File system is not NTFS. Skipped applying motw");
+            return;
+        }
+
+        // Attachment execution service needs STA to succeed, so we'll create a new thread and CoInitialize with STA.
+        auto updateMotw = [&]() -> HRESULT
+        {
+            Microsoft::WRL::ComPtr<IAttachmentExecute> attachmentExecute;
+            RETURN_IF_FAILED(CoCreateInstance(CLSID_AttachmentServices, nullptr, CLSCTX_INPROC_SERVER, IID_PPV_ARGS(&attachmentExecute)));
+            RETURN_IF_FAILED(attachmentExecute->SetLocalPath(filePath.c_str()));
+            RETURN_IF_FAILED(attachmentExecute->SetSource(Utility::ConvertToUTF16(source).c_str()));

```

Comment:
>SetSource [](start = 48, length = 9)

Does this end up saving the source string into the alternate stream?

Reply:
Sent to internal discussion

---
In reply to: [510501417](https://github.com/microsoft/winget-cli/pull/625#discussion_r510501417) [](ancestors = 510501417)